File: Algorithms/Merge Sort/merge_sort.py
import numpy as np


def merge(A, szA, B, szB):
    idxA = 0
    idxB = 0
    idxC = 0

    C = np.zeros(szA + szB, dtype=np.int32)

    while szA > idxA and szB > idxB:
        if A[idxA] > B[idxB]:
            C[idxC] = B[idxB]
            idxB += 1
        else:
            C[idxC] = A[idxA]
            idxA += 1
        idxC += 1

    while szA > idxA:
        C[idxC] = A[idxA]
        idxA += 1
        idxC += 1

    while szB > idxB:
        C[idxC] = B[idxB]
        idxB += 1
        idxC += 1

    A[: szA + szB] = C[: szA + szB]


def mergeSort(A, sz):
    if sz < 2:
        return

    half = sz // 2

    mergeSort(A, half)
    mergeSort(A[half:], sz - half)

    merge(A, half, A[half:], sz - half)

File: Algorithms/Merge Sort/mergeSort.cpp
#include <iostream>

using namespace std;

void merge(int *A, int szA, int *B, int szB)
{

    int idxA = 0;
    int idxB = 0;
    int idxC = 0;

    int *C = new int[szA + szB];

    while (szA > idxA && szB > idxB)
    {

        if (A[idxA] > B[idxB])
            C[idxC++] = B[idxB++];

        else
            C[idxC++] = A[idxA++];
    }

    while (szA > idxA)
        C[idxC++] = A[idxA++];

    while (szB > idxB)
        C[idxC++] = B[idxB++];

    for (size_t i = 0; i < szA + szB; i++)
    {
        A[i] = C[i];
    }

    delete[] C;
}

void mergeSort(int *A, int sz)
{

    if (sz < 2)
        return;

    int half = sz / 2;

    mergeSort(A, half);
    mergeSort(A + half, sz - half);

    merge(A, half, A + half, sz - half);
}
